### ISSUE TITLE
haskell_module fails when reexporting module definitions

### ISSUE THREAD

**Issue by facundominguez** [tweag/rules_haskell#1615]
The following test fails on 3cd1eed0.
```
$ nix-shell --pure --run "bazel build //tests/haskell_module/multiple:leaf"
INFO: Invocation ID: daabbbba-7c15-4eae-bfef-768469735989
INFO: Analyzed 5 targets (0 packages loaded, 0 targets configured).
INFO: Found 5 targets...
ERROR: /home/facundo/tweag/rules_haskell/tests/haskell_module/multiple/BUILD.bazel:29:15: HaskellBuildObject //tests/haskell_module/multiple:leaf failed: (Exit 1): ghc_wrapper failed: error executing command bazel-out/host/bin/haskell/ghc_wrapper bazel-out/k8-fastbuild/bin/tests/haskell_module/multiple/compile_flags_leaf_HaskellBuildObject ... (remaining 1 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox
Failed to load interface for ‘Root’
Use -v (or `:set -v` in ghci) to see a list of the files searched for.
INFO: Elapsed time: 0.879s, Critical Path: 0.75s
INFO: 3 processes: 2 internal, 1 linux-sandbox.
FAILED: Build did NOT complete successfully

```

`Root.hi` is missing from the inputs of the build action for `Leaf.hs`. Presumably, transitive dependencies on `.hi` files should be fed to the inputs of the build action.

**Comment by facundominguez** [comment]
In response to https://github.com/tweag/rules_haskell/pull/1616#issuecomment-947132827

> If Leaf.hs depends on the contents of Root.hs directly, should it list that dependency?

I don't think we can expect users to compute and maintain the transitive closure of dependencies for every module, nor is there much point in adding all of that information to the graph which is already expressed in its structure.

Asking `gazelle_haskell_modules` to infer whether a transitive dependency is necessary or not seems overkill as well. It would need to do the same name resolution that the compiler does in order to know which reexport comes from where. Additionally, there could be multiple conditions under which GHC might make a transitive dependency necessary even if not using reexports, maybe because of inlining. I don't think we would want to make the user mindful of all of that.

**Comment by facundominguez** [comment]
> Leaf should not be able to import Root without also declaring deps = [":Root", ...].

Ah, I think I finally get this. And I'm in trouble now. I don't know how to make the interface files available to GHC while at the same time not making the modules importable when they haven't been listed explicitly in `deps`.

**Comment by facundominguez** [comment]
Fixed in https://github.com/tweag/rules_haskell/pull/1623, https://github.com/tweag/rules_haskell/pull/1624, and https://github.com/tweag/rules_haskell/pull/1625.

We just made all transitive dependencies available, allowing to import a module in the transitive dependencies even if not listed in the direct dependencies. We reasoned that despite of departing from the bazel recommendations, the build failures we get are reproducible from every angle, and fixing it seems to involve a non-trivial effort.

**Comment by aherrmann** [comment]
> We reasoned that despite of departing from the bazel recommendations, the build failures we get are reproducible from every angle, and fixing it seems to involve a non-trivial effort.

I don't agree with that statement. As discussed before and as described [in the Bazel documentation](https://docs.bazel.build/versions/4.2.1/build-ref.html#actual_and_declared_dependencies), direct dependencies should be required to be declared as such. To unblock TH support we can accept this for now, but, I still consider this a bug that should be fixed going forward. I.e. as said before

>  [...] the point about implicit dependencies still stands:
> 
> > `Leaf` should not be able to `import Root` without also declaring `deps = [":Root", ...]`.

Could you open a dedicated issue to track this so we don't forget?

**Comment by facundominguez** [comment]
> Could you open a dedicated issue to track this so we don't forget?

I've created the new issue #1626.

**Comment by aherrmann** [comment]
Thank you!
